engine/src/cubos/engine/tools/voxel_palette_editor/plugin.cpp

```diff
@@ -0,0 +1,183 @@
+
+#include <cubos/engine/imgui/plugin.hpp>
+#include <cubos/engine/renderer/plugin.hpp>
+#include <cubos/engine/tools/asset_explorer/plugin.hpp>
+#include <cubos/engine/tools/voxel_palette_editor/plugin.hpp>
+#include <cubos/engine/voxels/plugin.hpp>
+
+using cubos::core::data::old::Debug;
+using cubos::core::ecs::EventReader;
+using cubos::core::ecs::Read;
+using cubos::core::ecs::Write;
+
+using namespace cubos::engine;
+
+using tools::AssetSelectedEvent;
+
+struct SelectedPaletteInfo
+{
+    Asset<VoxelPalette> asset;
+    VoxelPalette paletteCopy;
+    bool modified;
+    Asset<VoxelPalette> next;
+};
+
+static void savePaletteUiGuard(Write<Assets> assets, Write<SelectedPaletteInfo> selectedPalette)
+{
+    // The popup only shows when we've already opened a palette and want to show another one.
+    // Thus if we haven't set the next asset variable we can just stop here.
+    if (selectedPalette->next.isNull())
+    {
+        return;
+    }
+
+    ImVec2 center = ImGui::GetMainViewport()->GetCenter();
+    ImGui::SetNextWindowPos(center, ImGuiCond_Appearing, ImVec2(0.5F, 0.5F));
+
+    if (ImGui::BeginPopupModal("Save Palette?", nullptr, ImGuiWindowFlags_AlwaysAutoResize))
+    {
+        bool optionSelected = false;
+
+        ImGui::Text("Do you want to save the modified palette?");
+        ImGui::Separator();
+
+        if (ImGui::Button("Yes", ImVec2(120, 0)))
+        {
+            CUBOS_INFO("Saving palette asset {} modificaations", Debug(selectedPalette->next));
+            assets->store(selectedPalette->asset, selectedPalette->paletteCopy);
+            assets->save(selectedPalette->asset);
+            optionSelected = true;
+        }
+
+        ImGui::SetItemDefaultFocus();
+        ImGui::SameLine();
+
+        if (ImGui::Button("No", ImVec2(120, 0)))
+        {
+            CUBOS_DEBUG("Discarding palette asset {} modifications", Debug(selectedPalette->next));
+            optionSelected = true;
+        }
+
+        if (optionSelected)
+        {
+            selectedPalette->asset = assets->load(selectedPalette->next);
+            selectedPalette->paletteCopy = assets->read(selectedPalette->asset).get();
+            selectedPalette->modified = false;
+            ImGui::CloseCurrentPopup();
+        }
+
+        ImGui::EndPopup();
+    }
+}
+
+static void checkAssetEventSystem(EventReader<AssetSelectedEvent> reader, Write<Assets> assets,
+                                  Write<SelectedPaletteInfo> selectedPalette)
+{
+    for (const auto& event : reader)
+    {
+        if (assets->type(event.asset) == typeid(VoxelPalette))
+        {
+            CUBOS_INFO("Opening palette asset {}", Debug(event.asset));
+            if (!selectedPalette->asset.isNull() && selectedPalette->modified)
+            {
+                CUBOS_DEBUG("Opening save palette UI guard");
+                ImGui::OpenPopup("Save Palette?");
+                selectedPalette->next = event.asset;
+            }
+            else
```

Comment:
Should we have a different behaviour for if the user tries to open another asset while the guard is open? My intuition says in that case the new open instruction should be discarded. Ideally with some warning on the pop-up saying something like "You are already trying to open a palette!".
Any opinion, @RiscadoA ?